posts/blundell/You_have_permission_to.jade

```diff
@@ -0,0 +1,204 @@
+
+    With targetting Marshmallow, 6.0, API 23, the latest version of Android, strange things can happen if you don't get your permission flow correct. The code seems legit and yet it crashes every time. You have all the correct permissions in your AndroidManifest, you call the correct api to get your data. It runs perfectly well on your other emulator (_pre-marshmallow without realising_). Let's debug together.
+
+    Permissions are split into two categories, _dangerous_ and _normal_ the rest of this article is implicitly talking about _dangerous_ permissions.
+    You can read more about the [difference here.](http://developer.android.com/guide/topics/security/permissions.html#normal-dangerous) 
+
+    > "Once you eliminate the impossible, whatever remains, no matter how improbable, must be the truth."
+    >                                                           - Arthur Conan Doyle
+
+    As a good developer you've rechecked the code thrice, you've rebooted the emulator, hell you even closed Android Studio and re-opened it, _did you try Eclipse ... I hope you're not that desperate_! What else has changed... of course you're running against a new version of Android. 
+
+    Here is the simplest example to reproduce the discrepency between Lollipop and Marshmallow: 
+
+    MainActivity.java
+
+    ```java
+    import android.os.Bundle;
+    import android.provider.ContactsContract.Contacts;
+    import android.support.v7.app.AppCompatActivity;
+
+    public class MainActivity extends AppCompatActivity {
+        @Override
+        protected void onCreate(Bundle savedInstanceState) {
+            super.onCreate(savedInstanceState);
+            setContentView(R.layout.activity_main);
+
+            Contacts.getLookupUri(getContentResolver(), Contacts.CONTENT_URI);
+        }
+    }
+    ```
+
+    AndroidManifest.xml
+
+    ```xml
+    <?xml version="1.0" encoding="utf-8"?>
+    <manifest xmlns:android="http://schemas.android.com/apk/res/android"
+      package="com.blundell.contactscrash">
+
+      <uses-permission android:name="android.permission.READ_CONTACTS" />
+
+      <application
+        android:allowBackup="true"
+        android:icon="@mipmap/ic_launcher"
+        android:label="@string/app_name"
+        android:theme="@style/AppTheme">
+        <activity android:name=".MainActivity">
+          <intent-filter>
+            <action android:name="android.intent.action.MAIN" />
+
+            <category android:name="android.intent.category.LAUNCHER" />
+          </intent-filter>
+        </activity>
+      </application>
+
+    </manifest>
+    ```
+
+    Although this code will not do much, it still requires the system `READ_CONTACTS` permission to run, and on Lollipop it will run and your application will just sit their twiddling its thumbs. However on Marshmallow it will crash violently and your world will implode. What the hell, what happened to backward compatability??* Lets debug the stacktrace after that horrendous crash:
```

Comment:
`and on Lollipop it will run and your application will just sit their twiddling its thumbs. However on Marshmallow it will crash violently and your world will implode. What the hell, what happened to backward compatability??* Lets debug the stacktrace after that horrendous crash:`

->

`their` -> `there`

`compatability` -> `compatibility`

`Lets` -> `Let's`